adapters/amqp-vertx/src/main/java/org/eclipse/hono/adapter/amqp/impl/VertxBasedAmqpProtocolAdapter.java

```diff
@@ -683,42 +683,43 @@ private Span newSpan(final String operationName, final Device authenticatedDevic
-                                timer);
-                    } else {
-                        getTenantConfiguration(sourceAddress.getTenantId(), commandContext.getTracingContext())
-                                .compose(tenantObject -> checkMessageLimit(tenantObject, command.getPayloadSize()))
-                                .setHandler(result -> {
-                                    if (result.succeeded()) {
-                                        onCommandReceived(sender, commandContext);
-                                    } else {
-                                        // issue credit so that application(s) can send the next command
-                                        commandContext.reject(getErrorCondition(result.cause()), 1);
-                                        metrics.reportCommand(
-                                                command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
-                                                sourceAddress.getTenantId(),
-                                                ProcessingOutcome.from(result.cause()),
-                                                command.getPayloadSize(),
-                                                timer);
-                                    }
-                                });
-                    }
+                    final Future<TenantObject> tenantTracker = getTenantConfiguration(sourceAddress.getTenantId(),
+                            commandContext.getTracingContext());
+
+                    tenantTracker.compose(tenantObject -> {
+                        if (!command.isValid()) {
+                            return Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_BAD_REQUEST,
+                                    "malformed command message"));
+                        } else if (!sender.isOpen()) {
+                            commandContext.release(1);
+                            metrics.reportCommand(
+                                    command.isOneWay() ? Direction.ONE_WAY : Direction.REQUEST,
+                                    sourceAddress.getTenantId(),
+                                    tenantObject,
+                                    ProcessingOutcome.UNDELIVERABLE,
+                                    command.getPayloadSize(),
+                                    timer);
+                            return Future.succeededFuture();
+                        } else {
+                            return checkMessageLimit(tenantObject, command.getPayloadSize())
+                                    .compose(ok -> {
+                                        onCommandReceived(tenantObject, sender, commandContext);
+                                        return Future.succeededFuture();
+                                    });
+                        }
+                    }).setHandler(result -> {
+                        if (result.failed()) {
```

Comment:
it looks a little strange that there is only a *failed* block. What about moving the *onCommandReceived* invocation down here into a *succeeded* block and just leave the *checkMessageLimit* invocation in the block above?

Reply:
The `else if ` block comprising `!sender.isOpen()` returns _succeeded Future_. Therefore _onCommandReceived_  cannot be directly moved to the _succeeded_ block. Personally I would prefer a `otherwise` block instead of `setHandler` here. IMHO it also reduces no. of lines of code as there is no need for _if_ and _else_ blocks for _success_ or _failure_. Personally I find the `otherwise` block as intuitive as it deals only with _failures_. I guess its also a matter of taste. If you insist on using `setHandler`, I can refactor accordingly the whole part so that the _onCommandReceived_ can be moved to the _succeeded_ block.

Reply:
then I guess you should replace the *setHandler* with *otherwise*, right? :-)

Reply:
yes. :-)